exercises/building_suffix_arrays/descr.md

```diff
@@ -0,0 +1,543 @@
+              9
+              7
+              4
+              1
+              0
+              8
+              6
+              3
+              5
+              2
+```
+
+that is, in OCaml notation, the array `[|9; 7; 4; 1; 0; 8; 6; 3; 5; 2|]`.
+
+## Preliminaries
+
+In this exercise, we manipulate many arrays. Thus, we wish to refer to
+the most common operations on arrays via short names. Our prelude makes
+the following definitions:
+
+```
+  let copy = Array.copy
+  let init = Array.init
+  let iter = Array.iter
+  let length = Array.length
+  let make = Array.make
+  let map = Array.map
+```
+
+These functions are provided by the module
+[Array](https://caml.inria.fr/pub/docs/manual-ocaml/libref/Array.html)
+of the OCaml standard library.
+We also use several of the functions provided by the modules
+[String](https://caml.inria.fr/pub/docs/manual-ocaml/libref/String.html)
+and
+[List](https://caml.inria.fr/pub/docs/manual-ocaml/libref/List.html),
+but do not introduce short names for these functions.
+
+As announced earlier, a suffix of a string `s` is represented as an index into
+the string `s`, its start index. Thus, we make the following type definition:
+
+```
+  type suffix = int
+```
+
+While programming and debugging, it is convenient to have a function that
+converts a value of type `suffix` to an actual suffix, that is, a string.
+By mapping this function over a suffix array, we obtain a function that
+converts a suffix array to a human-readable form. Thus, our prelude makes
+the following definitions:
+
+```
+  let suffix (s : string) (i : suffix) : string =
+    let n = String.length s in
+    String.sub s i (n-i)
+
+  let suffixes (s : string) (a : suffix array) : string array =
+    map (suffix s) a
+```
+
+As an example, try evaluating the OCaml expression
+`suffixes "mississipi" [|9; 7; 4; 1; 0; 8; 6; 3; 5; 2|]`
+and check that you get the result that you expect.
+
+Because the function `suffixes` turns a suffix array of size `n` into a string
+array whose size is proportional to `n^2`, it should never be used as part of
+an efficient algorithm. It is meant to be used only while programming and
+debugging.
+
+Our prelude also defines a function that extracts a prefix out of a string:
+
+```
+  let prefix (h : int) (s : string) : string =
+    let n = String.length s in
+    if n <= h then s else String.sub s 0 h
+```
+
+## Recognizing a Suffix Array
+
+As a warm-up, and before attempting to build a suffix array, we first write
+code that checks whether an array `a` is the suffix array of the string `s`.
+
+**Question 1.** Write a function `is_sorted`
+of type `('a -> 'a -> bool) -> 'a array -> bool`
+such that `is_sorted (<=) a`
+determines whether the array `a` is sorted
+with respect to the preorder `(<=)`.
+
+**Question 2.** Write a function `is_permutation`
+of type `int array -> bool`
+such that `is_permutation p`
+determines whether the array `p` represents a valid
+permutation of the semi-open interval `[0,n)`,
```

Comment:
I know that some students do not always understand the natural encoding of permutations as arrays. Maybe add a short paragraph about this when you show the compact representation of suffix arrays a bit earlier?

Reply:
Good suggestion, done.